storage/replica_leader_lease.go

```diff
@@ -20,100 +20,164 @@
+		if nextLease.Replica.ReplicaID == nextLeader.ReplicaID {
+			// Join a pending request asking for the same replica to become leader.
+			p.llChans = append(p.llChans, llChan)
+			return llChan
+		}
+		llChan <- roachpb.NewErrorf("request for different replica in progress "+
+			"(requesting: %+v, in progress: %+v)",
+			nextLeader.ReplicaID, nextLease.Replica.ReplicaID)
 		return llChan
 	}
-	if r.store.IsDrainingLeadership() || !r.store.Stopper().RunAsyncTask(func() {
-		pErr := func() *roachpb.Error {
-			// TODO(tschottdorf): get duration from configuration, either as a
-			// config flag or, later, dynamically adjusted.
-			startStasis := timestamp.Add(int64(r.store.ctx.leaderLeaseActiveDuration), 0)
-			expiration := startStasis.Add(int64(r.store.Clock().MaxOffset()), 0)
-
-			// Prepare a Raft command to get a leader lease for this replica.
-			desc := r.Desc()
-			_, replica := desc.FindReplica(r.store.StoreID())
-			if replica == nil {
-				return roachpb.NewError(roachpb.NewRangeNotFoundError(r.RangeID))
-			}
-			args := &roachpb.LeaderLeaseRequest{
-				Span: roachpb.Span{
-					Key: desc.StartKey.AsRawKey(),
-				},
-				Lease: roachpb.Lease{
-					Start:       timestamp,
-					StartStasis: startStasis,
-					Expiration:  expiration,
-					Replica:     *replica,
-				},
-			}
-			ba := roachpb.BatchRequest{}
-			ba.Timestamp = r.store.Clock().Now()
-			ba.RangeID = r.RangeID
-			ba.Add(args)
-
-			// Send lease request directly to raft in order to skip unnecessary
-			// checks from normal request machinery, (e.g. the command queue).
-			// Note that the command itself isn't traced, but usually the caller
-			// waiting for the result has an active Trace.
-			ch, _, err := r.proposeRaftCommand(r.context(context.Background()), ba)
-			if err != nil {
-				return roachpb.NewError(err)
-			}
-
-			// If the command was committed, wait for the range to apply it.
-			select {
-			case c := <-ch:
-				if c.Err != nil {
-					if log.V(1) {
-						log.Infof("failed to acquire leader lease for replica %s: %s", r.store, c.Err)
-					}
-				}
-				return c.Err
-			case <-r.store.Stopper().ShouldDrain():
-				return roachpb.NewError(r.newNotLeaderError(nil, r.store.StoreID()))
-			}
-		}()
+	// No request in progress. Let's propose a LeaderLease command asynchronously.
+	// TODO(tschottdorf): get duration from configuration, either as a
+	// config flag or, later, dynamically adjusted.
+	startStasis := timestamp.Add(int64(replica.store.ctx.leaderLeaseActiveDuration), 0)
+	expiration := startStasis.Add(int64(replica.store.Clock().MaxOffset()), 0)
+	leaseReq := roachpb.LeaderLeaseRequest{
+		Span: roachpb.Span{
+			Key: startKey,
+		},
+		Lease: roachpb.Lease{
+			Start:       timestamp,
+			StartStasis: startStasis,
+			Expiration:  expiration,
+			Replica:     nextLeader,
+		},
+	}
+	if !replica.store.Stopper().RunAsyncTask(func() {
+		pErr := replica.executeLeaderLeaseCommand(leaseReq, timestamp, startKey)
```

Comment:
I think you should pass `leaseReq.Lease` instead, then you can construct `leaseReq` in that method (and you wouldn't have to pass the timestamp).